rfc/20241118-module-vars-and-outputs-deprecation.md

```diff
@@ -0,0 +1,148 @@
+configuration.
+
+> [!NOTE]
+> The result of `tofu validate`, `tofu plan`, `tofu apply` and `tofu test` should include deprecation warnings.
+
+### User Documentation
+
+Module authors should be able to use a new field named `deprecated` in both `variable` and `output` blocks. This field 
+should accept a required string value for a warning message:
+
+```hcl
+variable "this_is_my_variable" {
+  type        = string
+  description = "This is a variable for the old way of configuring things."
+  deprecated  = "This variable will be removed on 2024-12-31. Use that_is_my_variable instead."
+}
+
+output "this_is_my_output" {
+  value       = some_resource.some_name.some_field
+  description = "This is an output for the old way of using things."
+  deprecated  = "This output will be removed on 2024-12-31. Use that_is_my_output instead."
+}
+```
+
+#### Deprecated module variable warning
+
+If module caller specify non-null value for a deprecated variable, OpenTofu shoould produce a deprecation warning as 
+specified by module author:
+
+```hcl
+│ Warning: The variable "this_is_my_variable" is marked as deprecated by module author.
+│ 
+│   on mod/main.tf line 9, in module call "mod":
+│    9:     this_is_my_variable = "something"
+│ 
+│ This variable will be removed on 2024-12-31. Use that_is_my_variable instead.
+```
+
+> [!NOTE]
+> It is possible for module user to specify a null value for a deprecated variable and receive no warnings. OpenTofu treats
+> variables with explicit null values the same way as if it was ommited completely. This is not an ideal, however an idiomatic
+> OpenTofu language approach. See `null` values note [in the docs](https://opentofu.org/docs/language/expressions/types/#types).
+
+#### Deprecated module output warning
+
+TBD
+
+### Technical Approach
+
+TBD
+
+### Open Questions
+
+* Do we want to support silencing of deprecation warnings?
+
+### Future Considerations
+
+It is hard to implement generic deprecation mechanism for the OpenTofu language. However, this solution should be generic 
+enough from the UX point of view to potentially be extended for other purposes. This way we keep consistent experience for
+OpenTofu users.
+
+Also, we want to keep compatibility with Terraform's deprecation mechanisms.
+
+## Potential Alternatives
+
+Here is the list of other potential options (mostly from the UX perspective), which were considered at the time of writing this RFC.
+
+#### HCL comment
+
+Right now, OpenTofu doesn't treat comments as something with a special meaning so this feature would introduce a whole new set
+of functionality to be used across the OpenTofu. It would require a separate RFC to define how this notion could potentially
+evolve.
+
+```hcl
+# @deprecated: This variable will be removed on 2024-12-31. Use that_is_my_variable instead.
+variable "this_is_my_variable" {
+  type = string
+  description = "This is a variable for the old way of configuring things."
+}
+```
```

Comment:
The main `hclsyntax` parser does not currently retain comments at all.

There is some basic support for retaining comments appearing immediately before a block or argument (or on the same line as a one-line argument) in [`hclwrite`](https://pkg.go.dev/github.com/hashicorp/hcl/v2@v2.23.0/hclwrite) but that's primarily just so that a doc comment associated with something will get deleted if that something is deleted, and not to allow general-purpose "programming with comments".